rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,337 @@
+to the GOV.UK Account system to do the actual authentication, but we
+need something on www.gov.uk itself to set the cookie.
+
+### Non-personalised parts of GOV.UK need to manipulate the session cookie too
+
+It's unlikely that personalisation will roll out to all of GOV.UK in
+one go.  There will be islands of personalised content.  Currently
+there is the Transition Checker.  Maybe next will be some guidance
+pages, or something else.  We want to keep the user's session alive
+while they are browsing the non-personalised parts of GOV.UK,
+otherwise we risk this bad experience:
+
+1. The user logs in to use some personalised part of GOV.UK.
+
+2. The user then spends spends 30 minutes (or whatever we use for the
+   session duration) viewing non-personalised parts of GOV.UK, but
+   without leaving the site.
+
+3. The user then tries to use another personalised part of GOV.UK, but
+   their session has expired, because the non-personalised parts
+   weren't bumping the expiration time on every page view.
+
+### We can't cache as effectively
+
+The Fastly docs have some [comments on the risks of cookies][fastly-cookies-risks]:
+
+> Cookies can lead to undesirable outcomes.  At worst, if a cookie
+> header is forwarded to your backend server, the backend uses the
+> cookie value to generate personalized content, and that content is
+> then cached by Fastly, a user may end up receiving content intended
+> for someone else.  A theoretical solution to this, adding a Vary:
+> Cookie header to the response, leads to another bad outcome: the
+> response is most likely not cacheable at all, and Fastly will
+> forward all requests to your backend.
+
+[fastly-cookies-risks]: https://developer.fastly.com/reference/http-headers/Cookie/
+
+## Proposal
+
+### Manage the cookie entirely in VCL
+
+[Fastly's best practices for cookies][fastly-cookies-practices]
+suggest parsing cookies into custom headers and using these headers
+for caching purposes, rather than caching based on the entire
+Set-Cookie header (which in our case also contains non-account-related
+things like A/B test bucket assignment, Google Analytics session ID,
+and cookie consent preferences).
+
+Rather than have multiple GOV.UK frontend apps worry about receiving
+the cookie and bumping its expiration time, if we manage the cookie
+entirely in VCL—parsing it into a new `GOVUK-Session-ID` header—we
+only need to worry about returning appropriate response headers.
+
+We will need to make three changes to our VCL.
+
+[fastly-cookies-practices]: https://developer.fastly.com/reference/http-headers/Cookie/#best-practices
+
+#### Preventing tampering with the cookie
+
+We can't rely on client-side cookie timeouts to log users out.  If an
+attacker gets a copy of a user's session cookie, they can include it
+in any requests they send, regardless of whether it has expired.  So
+we need to sign the cookie value and expiration time to prevent
+tampering.
```

Comment:
Will apps do any validation of the session header(s) at all? Or will they just receive a Session-ID header and trust that this is valid, similar to bearer token authentication?

I guess if we have these concerns about an attacker getting hold of a user's session cookie then we have to consider can they do anything with the Session-ID alone (which can be determined from the cookie)? I guess at a basic level something needs to make sure Fastly VCL always has a `unset req.http.GOVUK-Session-ID;` which isn't behind conditionals so an attacker can't just set the header manually. Then beyond that we have edge cases like the `www-origin` entry (which I believe is just for VPN access) where we'd not have this cookie verification.

Reply:
I was thinking of having the app trust the session ID header, as if it's only set by Fastly (which verifies the signature) then we're not opening up any vulnerabilities.  Though as you spotted I missed a place where the header needed to be unset.

An alternative would be passing the signed cookie value through in a header, and doing validation in the app which consumes it.  That means we'd need to keep the secret in sync between Fastly (so we can continue to bump the cookie expiration time through VCL) and puppet (so our app can validate the header).  What do you think?

Reply:
> An alternative would be passing the signed cookie value through in a header, and doing validation in the app which consumes it. That means we'd need to keep the secret in sync between Fastly (so we can continue to bump the cookie expiration time through VCL) and puppet (so our app can validate the header). What do you think?

Yup that seems a reasonable solution and would resolve the problem. I can see that it'd be a bit of a pain to manage so might be worth validating the necessity first. 

As I understand the worry is that the way an account could be hijacked through discovering a cookie (is that the only public availability of this session id?) and we're accepting that by discovering that cookie they can immediately use the account regardless of verification. The security feature kicks in after that cookie has aged, so if we're in a world where session id lives indefinitely and a cookie lifetime is short (30m) then it offers a security advantage of limiting the window of a bad actor. Whereas if we were in a world where the cookie had much longer life time (1 week) and a session id only lasted for 24 hours then it'd be an unnecessary feature as it wouldn't change any risk.

Perhaps it's worth considering a think about cookie lifetime (since I understand that's an open question), session id lifetime and how secret session id is?

Reply:
@OllieJC - do you have any thoughts on this?  To summarise, we're in this situation:

- A user is assigned a random session ID when logging in.
- This session ID is stored in a (secure, httponly) cookie with a short (undecided - 30 minute?) expiration time.
- We sign the cookie value + expiration time so that an attacker can't tamper with it.
- On every request, we bump up the expiration time.

The problem is that the session ID is kind of sensitive.

We *don't* regularly change the session ID, as we don't want to hit our origin on every request (GOV.UK can't cope with every request going to origin).  Which means that if an attacker learns the session ID, they can access the account it corresponds to if they can:

- forge a cookie (which would require being able to sign it with out secret), or
- bypass the cookie mechanism and pass the session ID to our origin directly.

---

We could put an expiration time on the session ID, so that a user would need to log in again even if they had been continuously browsing GOV.UK.  So this would be an upper limit on session duration.  But we'd probably want a long upper limit (days) to avoid being annoying.

We could pass the full signed cookie value and not just the session ID to our origin, so an attacker would need to forge a cookie (which requires knowing a secret).  But we'll need to keep the signing secret consistent between our apps and our Fastly config.

Reply:
I've implemented Ollie's suggestions now.  I've kept the custom headers, because I think our nginx config which strips out `Set-Cookie` *does* add some value (eg, it ensures we can't make a mistake and accidentally leak cookies from an app, adding cookies has to be considered).